### ISSUE TITLE
when to use map() or apply() for Function<T, R> interface?

### ISSUE THREAD

**Issue by jjiehh** [nus-cs2030/2324-s1#88]
Hi, could anyone explain as I am confused when to use these methods for Function<T, R>?

**Comment by jr-kong** [comment]
i think map is usually associated with java streams, where it is more convenient to transform elements in a stream using the function. one example is:

List<Integer> numbers = List.<Integer>of(1, 2, 3, 4, 5);
List<Integer> doubledNumbers = numbers.stream()
    .map(n -> n * 2)
    .collect(Collectors.toList());

where the elements in the list are being put in a stream and mapped

then for apply is just usually used when u want to apply a function to a specific input:
Function<Integer, Integer> func = n -> n * 2;
int result = func.apply(3);  => 6

**Comment by DarkAxel01** [comment]
idk if is right to say this, but when the lambda function is use in a way like a method, u use .apply(), and for other things such as map,flatmap etc. the lambda function will be the input to these methods.

**Comment by saddle196883** [comment]
``map()`` is a method that you **use** in a functor (and Monad, but every Monad is a functor.), when you want to use a function ``f`` to transform the encapsulated value in the functor into a new value, you use the ``.map(f)`` method to ask the functor to do the transformation for you. At no point here will you be using ``f.apply(thing)``, because the functor class is supposed to do the applying for you.

On the other hand, if you are **implementing** the functor, then you will be applying the given function for the client. In this case, there will usually be a call to ``f.apply(thing)`` in the method implementation of the ``map`` method.

**Comment by PVT-LF** [comment]
To put it simply, in most cases you simply take a context which has a `map()` method and you use the `Function` as an input for `map()`, as you're basically using the `Function` to tell the context **how** the mapping is supposed to happen.

Meanwhile `apply()` is a method of `Function` and takes in a `T`, and returns an `R`.

TLDR: `map()` is a method of the context while `apply()` is a method of the `Function`, and what you input and what you get as output is fundamentally different. Knowing the difference is the first step in obtaining compilable code.
